### PR TITLE
feat: support http type resource

### PR DESCRIPTION
**Pull request by zhujian7**
**What this PR does / why we need it**:

- Workflow Engine supports a new type resource: `Http`，you can use this feature to pull/push resources from/to a remote server.
- Cyclone Server provides APIs to receive and send artifacts. 

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #

Reference to #

**Special notes for your reviewer**:

/cc @cd1989 @supereagle 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by zhujian7** [comment]
@supereagle PTAL

File: pkg/server/handler/v1alpha1/workflowrun.go [hunk 1]
```diff
@@ -459,3 +466,150 @@ func GetContainerLogs(ctx context.Context, project, workflow, workflowrun, tenan
...
+
+		files, err := ioutil.ReadDir(artifactFolder)
+		if err != nil {
+			log.Infof("Readdir %s error:%v", artifactFolder, err)
+			return nil, err
+		}
+
+		for _, file := range files {
+			// Only support display files
+			if file.IsDir() {
```
Comment (supereagle) [inline pkg/server/handler/v1alpha1/workflowrun.go]: It will be a little confused if we collect folders but not display them.

File: pkg/server/handler/v1alpha1/workflowrun.go [hunk 2]
```diff
@@ -459,3 +466,150 @@ func GetContainerLogs(ctx context.Context, project, workflow, workflowrun, tenan
 
 	return folderReader, headers, nil
 }
+
+// ReceiveStageArtifacts receives artifacts produced by workflowrun stage.
+func ReceiveStageArtifacts(ctx context.Context, workflowrun, namespace, stage string) error {
```
Comment (supereagle) [inline pkg/server/handler/v1alpha1/workflowrun.go]: Keep consistent name style with other functions?

```suggestion
func ReceiveArtifacts(ctx context.Context, workflowrun, namespace, stage string) error {
```

File: pkg/server/apis/v1alpha1/types.go [hunk 3]
```diff
@@ -345,3 +347,12 @@ const (
 	// CacheCleanupSucceeded ...
 	CacheCleanupSucceeded CacheCleanupPhase = "Succeeded"
 )
+
+// StageArtifact describes artifacts produced by a certain stage
+type StageArtifact struct {
+	// Stage name
+	Stage string `json:"stage"`
+	// Files name
+	File              string    `json:"file"`
```
Comment (supereagle) [inline pkg/server/apis/v1alpha1/types.go]: When there are many files, the stage name will be duplicated returned.
```suggestion
// StageArtifact describes artifacts produced by a certain stage
type StageArtifacts struct {
	// Stage name
	Stage string `json:"stage"`
	// Files name
	Files              []StageArtifactFile    `json:"files"`
}

type StageArtifactFile struct {
	// Files name
	File              string    `json:"file"`
	CreationTimestamp time.Time `json:"creationTimestamp"`
}
```

File: build/resolver/http/entrypoint.sh [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+      Supported commands are:
+      - help Print out this help messages.
+      - pull Send a http request to get sources to ${WORKDIR}/data, (Not implemented yet).
+      - push Compress files in ${WORKDIR}/data to a file named $COMPRESS_FILE_NAME and 
+        send a http request to upload the compressed file with form key $FORM_FILE_KEY 
+        to a remote server.
+        deep dive in implementation with pseudocode:
+          $ tar -cvf $COMPRESS_FILE_NAME ${WORKDIR}/data
+          $ curl $URL -H $HEADERS -X $METHOD -F "$FORM_FILE_KEY@$COMPRESS_FILE_NAME" $CURL_EXTENSION
```
Comment (supereagle) [inline build/resolver/http/entrypoint.sh]:
```suggestion
          $ curl $URL -H $HEADERS -X $METHOD -F "$FORM_FILE_KEY=@$COMPRESS_FILE_NAME" $CURL_EXTENSION
```

File: pkg/server/apis/v1alpha1/types.go (hunk 3 above)
Comment (zhujian7) [reply pkg/server/apis/v1alpha1/types.go]: The notes of `StageArtifact` is not correct, should be `StageArtifact describes artifacts produced by a workflowRun`

File: pkg/server/handler/v1alpha1/workflowrun.go (hunk 1 above)
Comment (zhujian7) [reply pkg/server/handler/v1alpha1/workflowrun.go]: we can not collect folders, all files will be compressed to a tar file.

**Comment by zhujian7** [comment]
@supereagle  PTAL

File: pkg/server/handler/v1alpha1/workflowrun.go [hunk 5]
```diff
@@ -459,3 +466,150 @@ func GetContainerLogs(ctx context.Context, project, workflow, workflowrun, tenan
...
+		}
+	}
+
+	return types.NewListResponse(len(artifacts), artifacts), nil
+}
+
+// DownloadArtifact handles the request to download a artifact of a stage produced by a workflowRun.
+func DownloadArtifact(ctx context.Context, project, workflow, workflowrun, artifact, tenant, stage string, download bool) (io.ReadCloser, map[string]string, error) {
+	headers := make(map[string]string)
+	if download {
```
Comment (supereagle) [inline pkg/server/handler/v1alpha1/workflowrun.go]: Submit another PR to remove the `download` parameter.

Comment (zhujian7) [reply pkg/server/handler/v1alpha1/workflowrun.go]: https://github.com/caicloud/cyclone/pull/1398
